src/components/structures/RoomView.js

```diff
@@ -496,64 +493,72 @@ module.exports = React.createClass({
+                ]
+            };
+        }
+
+        if (DEBUG_SCROLL) console.log("sending search request");
+
+        var searchPromise = MatrixClientPeg.get().searchRoomEvents(
+            { filter: filter,
+              term: term,
+            });
+        this._handleSearchResult(searchPromise)
+            .done();
     },
 
-    // fire off a request for a batch of search results
-    _getSearchBatch: function(term, scope) {
-        this.setState({
-            searchInProgress: true,
-        });
+    _backPaginateSearch: function() {
+        if (DEBUG_SCROLL) console.log("sending search back-paginate request");
 
-        // make sure that we don't end up merging results from
-        // different searches by keeping a unique id.
-        //
-        // todo: should cancel any previous search requests.
-        var searchId = this.searchId = new Date().getTime();
+        var searchPromise = MatrixClientPeg.get().backPaginateRoomEventsSearch(
+            this.state.searchResults);
+        this._handleSearchResult(searchPromise)
+            .done();
+    },
 
+    _handleSearchResult: function(searchPromise) {
         var self = this;
+        var searchId = this.searchId;
 
-        if (DEBUG_SCROLL) console.log("sending search request");
-        MatrixClientPeg.get().search({ body: this._getSearchCondition(term, scope),
-                                       next_batch: this.nextSearchBatch })
-        .then(function(data) {
+        this.setState({
+            searchInProgress: true,
+        });
+
+        return searchPromise.then(function(results) {
             if (DEBUG_SCROLL) console.log("search complete");
             if (!self.state.searching || self.searchId != searchId) {
                 console.error("Discarding stale search results");
                 return;
             }
 
-            var results = data.search_categories.room_events;
-
             // postgres on synapse returns us precise details of the
             // strings which actually got matched for highlighting.
 
-            // combine the highlight list with our existing list; build an object
-            // to avoid O(N^2) fail
-            var highlights = {};
-            results.highlights.forEach(function(hl) { highlights[hl] = 1; });
-            self.state.searchHighlights.forEach(function(hl) { highlights[hl] = 1; });
-
-            // turn it back into an ordered list. For overlapping highlights,
+            // For overlapping highlights,
             // favour longer (more specific) terms first
-            highlights = Object.keys(highlights).sort(function(a, b) { b.length - a.length });
+            var highlights = results.highlights.sort(function(a, b) { b.length - a.length });
 
             // sqlite doesn't give us any highlights, so just try to highlight the literal search term
             if (highlights.length == 0) {
-                highlights = [ term ];
+                highlights = [ self.state.searchTerm ];
```

Comment:
We presumably *always* want to highlight the search term, so I would expect this to be checked even if there are `highlights`.